File: base/src/main/java/com/redefine/welike/base/util/DateTimeUtil.java
package com.redefine.welike.base.util;

import android.annotation.SuppressLint;

import com.redefine.welike.base.resource.ResourceTool;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.concurrent.TimeUnit;

/**
 * Created by liwb on 2018/1/11.
 * <p>
 * change by zhhl on 2018、7、3
 */

public class DateTimeUtil {
    public static final long SECOND_60 = 60 * 1000;
    public static final long HOUR_1 = 60 * 60 * 1000;
    public static final long MINUTE_1 = 60 * 1000;
    public static final long DAY_1 = 60 * 60 * 24 * 1000;

    private static SimpleDateFormat hiDateTimeFormatter = new SimpleDateFormat("dd-MM-yyyy H:mm");
    private static SimpleDateFormat hiDateFormatter = new SimpleDateFormat("dd-MM-yyyy");
//    private static SimpleDateFormat hiTimeFormatter = new SimpleDateFormat("dd-MM H:mm");
//    private static SimpleDateFormat hiDayFormatter = new SimpleDateFormat("H:mm");

    public static String formatDateTime(long time) {
        return hiDateFormatter.format(new Date(time));
    }

    public static String formatLogDateTime(long time) {
        return hiDateTimeFormatter.format(new Date(time));
    }

    public static String formatArticleDateTime(long time) {
        return hiDateTimeFormatter.format(new Date(time));
    }

    public static String formatModifyDate(long date) {
        return hiDateFormatter.format(date);
    }

    public static String formatImChatTime(long time) {
        long currentTime = System.currentTimeMillis();
        long offset = currentTime - time;

        Calendar currentDate = Calendar.getInstance();
        currentDate.setTimeInMillis(currentTime);

        Calendar receiveDate = Calendar.getInstance();
        receiveDate.setTimeInMillis(time);

        if (offset < SECOND_60) {
            return ResourceTool.getString(ResourceTool.ResourceFileEnum.FEED, "now");
        } else if (offset < HOUR_1) {
            return String.valueOf(offset / MINUTE_1) + " Min";
        } else if (offset < DAY_1) {
            return String.valueOf(offset / HOUR_1) + " Hrs";
        } else if (offset < 30 * DAY_1) {
            return String.valueOf(offset / DAY_1) + ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "day_Days");
        } else if (currentDate.get(Calendar.YEAR) == receiveDate.get(Calendar.YEAR)) {
            return getDateInYear(receiveDate);
        } else {
            return getDateOutYear(receiveDate);
        }
    }

    public static String formatPostPublishTime(long time) {
        return getCommonTime(time);
    }

    public static String getMessageReceiveTime(long time) {

        return getCommonTime(time);
    }

    private static String getCommonTime(long time) {
        long currentTime = System.currentTimeMillis();
        long offset = currentTime - time;
        Calendar currentDate = Calendar.getInstance();
        currentDate.setTimeInMillis(currentTime);
        Calendar postDate = Calendar.getInstance();
        postDate.setTimeInMillis(time);
        if (offset < SECOND_60) {
            return ResourceTool.getString(ResourceTool.ResourceFileEnum.FEED, "now");
        } else if (offset < HOUR_1) {
            return String.valueOf(offset / MINUTE_1) + " Min";
        } else if (offset < DAY_1) {
            return String.valueOf(offset / HOUR_1) + " Hrs";
        } else if (offset < 2 * DAY_1) {
            return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "day_yesterday");
        } else if (offset < 7 * DAY_1) {
            return getDay4Week(postDate);
        } else if (currentDate.get(Calendar.YEAR) == postDate.get(Calendar.YEAR)) {
            return getDateInYear(postDate);
        } else {
            return getDateOutYear(postDate);
        }
    }

    public static String secToTime(int time) {
        String timeStr = null;
        int hour = 0;
        int minute = 0;
        int second = 0;
        if (time <= 0)
            return "00:00";
        else {
            minute = time / 60;
            if (minute < 60) {
                timeStr = unitFormat(minute) + "min";
            } else {
                hour = minute / 60;
                if (hour > 99)
                    return "99:59:59";
                minute = minute % 60;
                if (minute == 0) {
                    timeStr = unitFormat(hour) + "h";

                } else {
                    timeStr = unitFormat(hour) + "h" + " " + unitFormat(minute) + "min";

                }
            }
        }
        return timeStr;
    }

    public static String unitFormat(int i) {
        String retStr = null;
        if (i >= 0 && i < 10)
            retStr = Integer.toString(i);
        else
            retStr = "" + i;
        return retStr;
    }

    public static String getVoteExpiredDayTime(long time) {

        boolean isAddDay = getRealVoteExpiredHourTime(time).equals("23");

        if (time > DAY_1) {
            return String.valueOf(time / DAY_1 + (isAddDay ? 1 : 0));
        } else {
            return isAddDay ? "1" : "0";
        }
    }

    private static String getRealVoteExpiredHourTime(long time) {
        if (time > HOUR_1) {
            return String.valueOf(time / HOUR_1 % 24);
        } else {
            return "0";
        }
    }


    public static String getVoteExpiredHourTime(long time) {
        long hours = time / HOUR_1 % 24;

        if (hours == 23) return "0";
        else return String.valueOf(hours + 1);
    }

    /**
     * MS turn every minute
     *
     * @param duration Millisecond
     * @return Every minute
     */
    @SuppressLint("DefaultLocale")
    public static String timeParse(long duration) {
        return String.format("%02d:%02d",
                TimeUnit.MILLISECONDS.toMinutes(duration),
                TimeUnit.MILLISECONDS.toSeconds(duration) -
                        TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(duration))
        );
    }

    /***
     * @return DAY 4 WEEK
     *
     */
    private static String getDay4Week(Calendar calendar) {

        int week = calendar.get(Calendar.DAY_OF_WEEK);

        switch (week) {
            case Calendar.MONDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_monday");
            case Calendar.TUESDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_tueday");
            case Calendar.WEDNESDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_wedday");
            case Calendar.THURSDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_thurday");
            case Calendar.FRIDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_friday");
            case Calendar.SATURDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_satday");
            case Calendar.SUNDAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "week_sunday");
            default:
                return "";
        }
    }


    /**
     * @return "22 Oct 2016 "
     */
    private static String getDateOutYear(Calendar calendar) {

        StringBuilder date = new StringBuilder();
        date.append(calendar.get(Calendar.DAY_OF_MONTH));
        date.append(' ');
        date.append(getDateMonth(calendar));
        date.append(' ');
        date.append(calendar.get(Calendar.YEAR));

        return date.toString();

    }

    /**
     * @return "22 Oct"
     */
    private static String getDateInYear(Calendar calendar) {

        StringBuilder date = new StringBuilder();
        date.append(calendar.get(Calendar.DAY_OF_MONTH));
        date.append(' ');
        date.append(getDateMonth(calendar));
        return date.toString();

    }

    /**
     * @return "Oct"
     */

    private static String getDateMonth(Calendar calendar) {

        int week = calendar.get(Calendar.MONTH);

        switch (week) {
            case Calendar.JANUARY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Jan");
            case Calendar.FEBRUARY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Feb");
            case Calendar.MARCH:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Mar");
            case Calendar.APRIL:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Apr");
            case Calendar.MAY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_May");
            case Calendar.JUNE:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_June");
            case Calendar.JULY:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_July");
            case Calendar.AUGUST:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Aug");
            case Calendar.SEPTEMBER:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Sep");
            case Calendar.OCTOBER:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Oct");
            case Calendar.NOVEMBER:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Nov");
            case Calendar.DECEMBER:
                return ResourceTool.getString(ResourceTool.ResourceFileEnum.COMMON, "mon_Dec");
            default:
                return "";
        }

    }


    public static boolean isSameDay(long lastTime, long currentTime) {

        Calendar currentDate = Calendar.getInstance();
        currentDate.setTimeInMillis(currentTime);

        Calendar lastTimeDate = Calendar.getInstance();
        lastTimeDate.setTimeInMillis(lastTime);

        return lastTimeDate.get(Calendar.DAY_OF_YEAR) == currentDate.get(Calendar.DAY_OF_YEAR);

    }

    public static boolean isOver5Min(long newTime, long preTime) {
        long diff = Math.abs(newTime - preTime);
        return diff > 5 * 60 * 1000;
    }


}

File: app/src/main/java/com/redefine/welike/common/util/DateTimeUtil.kt
package com.redefine.welike.common.util

import android.content.res.Resources
import com.redefine.welike.R
import java.text.SimpleDateFormat
import java.util.*

/**
 *
 * Name: DateTimeUtil
 * Author: weber
 * Email:
 * Comment: //TODO
 * Date: 2019-02-20 17:23
 *
 */

object DateTimeUtil {

    const val SECOND_60 = (60 * 1000).toLong()
    const val HOUR_1 = (60 * 60 * 1000).toLong()
    const val MINUTE_1 = (60 * 1000).toLong()
    const val DAY_1 = (60 * 60 * 24 * 1000).toLong()

    private val hiDateTimeFormatter = SimpleDateFormat("dd-MM-yyyy H:mm")
    private val hiDateFormatter = SimpleDateFormat("dd-MM-yyyy")
//    private static SimpleDateFormat hiTimeFormatter = new SimpleDateFormat("dd-MM H:mm");
//    private static SimpleDateFormat hiDayFormatter = new SimpleDateFormat("H:mm");

    fun formatDateTime(time: Long): String {
        return hiDateFormatter.format(Date(time))
    }

    fun formatLogDateTime(time: Long): String {
        return hiDateTimeFormatter.format(Date(time))
    }

    fun formatArticleDateTime(time: Long): String {
        return hiDateTimeFormatter.format(Date(time))
    }

    fun formatModifyDate(date: Long): String {
        return hiDateFormatter.format(date)
    }

    fun formatImChatTime(resources: Resources, time: Long): String {
        val currentTime = System.currentTimeMillis()
        val offset = currentTime - time

        val currentDate = Calendar.getInstance()
        currentDate.timeInMillis = currentTime

        val receiveDate = Calendar.getInstance()
        receiveDate.timeInMillis = time

        return when {
            offset < SECOND_60 -> resources.getString(R.string.now)
            offset < HOUR_1 -> (offset / MINUTE_1).toString() + " Min"
            offset < DAY_1 -> (offset / HOUR_1).toString() + " Hrs"
            offset < 30 * DAY_1 -> (offset / DAY_1).toString() + resources.getString(R.string.day_Days)
            currentDate.get(Calendar.YEAR) == receiveDate.get(Calendar.YEAR) -> getDateInYear(resources, receiveDate)
            else -> getDateOutYear(resources, receiveDate)
        }
    }

    fun formatPostPublishTime(resources: Resources, time: Long): String {
        return getCommonTime(resources, time)
    }

    fun getMessageReceiveTime(resources: Resources, time: Long): String {

        return getCommonTime(resources, time)
    }

    private fun getCommonTime(resources: Resources, time: Long): String {
        val currentTime = System.currentTimeMillis()
        val offset = currentTime - time
        val currentDate = Calendar.getInstance()
        currentDate.timeInMillis = currentTime
        val postDate = Calendar.getInstance()
        postDate.timeInMillis = time
        return when {
            offset < SECOND_60 -> resources.getString(R.string.now)
            offset < HOUR_1 -> (offset / MINUTE_1).toString() + " Min"
            offset < DAY_1 -> (offset / HOUR_1).toString() + " Hrs"
            offset < 2 * DAY_1 -> resources.getString(R.string.day_yesterday)
            offset < 7 * DAY_1 -> getDay4Week(resources, postDate)
            currentDate.get(Calendar.YEAR) == postDate.get(Calendar.YEAR) -> getDateInYear(resources, postDate)
            else -> getDateOutYear(resources, postDate)
        }
    }

    fun secToTime(time: Int): String {
        var timeStr: String? = null
        var hour = 0
        var minute = 0
        val second = 0
        if (time <= 0)
            return "00:00"
        else {
            minute = time / 60
            if (minute < 60) {
                timeStr = unitFormat(minute) + "min"
            } else {
                hour = minute / 60
                if (hour > 99)
                    return "99:59:59"
                minute %= 60
                timeStr = if (minute == 0) {
                    unitFormat(hour) + "h"

                } else {
                    unitFormat(hour) + "h" + " " + unitFormat(minute) + "min"

                }
            }
        }
        return timeStr
    }

    private fun unitFormat(i: Int): String {
        return if (i in 0..9)
            Integer.toString(i)
        else
            "" + i
    }

    fun getVoteExpiredDayTime(time: Long): String {

        val isAddDay = getRealVoteExpiredHourTime(time) == "23"

        return if (time > DAY_1) {
            (time / DAY_1 + if (isAddDay) 1 else 0).toString()
        } else {
            if (isAddDay) "1" else "0"
        }
    }

    private fun getRealVoteExpiredHourTime(time: Long): String {
        return if (time > HOUR_1) {
            (time / HOUR_1 % 24).toString()
        } else {
            "0"
        }
    }


    fun getVoteExpiredHourTime(time: Long): String {
        val hours = time / HOUR_1 % 24

        return if (hours == 23L)
            "0"
        else
            (hours + 1).toString()
    }

    /***
     * @return DAY 4 WEEK
     */
    private fun getDay4Week(resources: Resources, calendar: Calendar): String {

        val week = calendar.get(Calendar.DAY_OF_WEEK)

        return when (week) {
            Calendar.MONDAY -> resources.getString(R.string.week_monday)
            Calendar.TUESDAY -> resources.getString(R.string.week_tueday)
            Calendar.WEDNESDAY -> resources.getString(R.string.week_wedday)
            Calendar.THURSDAY -> resources.getString(R.string.week_thurday)
            Calendar.FRIDAY -> resources.getString(R.string.week_friday)
            Calendar.SATURDAY -> resources.getString(R.string.week_satday)
            Calendar.SUNDAY -> resources.getString(R.string.week_sunday)
            else -> ""
        }
    }


    /**
     * @return "22 Oct 2016 "
     */
    private fun getDateOutYear(resources: Resources, calendar: Calendar): String {

        val date = StringBuilder()
        date.append(calendar.get(Calendar.DAY_OF_MONTH))
        date.append(' ')
        date.append(getDateMonth(resources, calendar))
        date.append(' ')
        date.append(calendar.get(Calendar.YEAR))

        return date.toString()

    }

    /**
     * @return "22 Oct"
     */
    private fun getDateInYear(resources: Resources, calendar: Calendar): String {

        val date = StringBuilder()
        date.append(calendar.get(Calendar.DAY_OF_MONTH))
        date.append(' ')
        date.append(getDateMonth(resources, calendar))
        return date.toString()

    }

    /**
     * @return "Oct"
     */

    private fun getDateMonth(resources: Resources, calendar: Calendar): String {

        val week = calendar.get(Calendar.MONTH)

        return when (week) {
            Calendar.JANUARY -> resources.getString(R.string.mon_Jan)
            Calendar.FEBRUARY -> resources.getString(R.string.mon_Feb)
            Calendar.MARCH -> resources.getString(R.string.mon_Mar)
            Calendar.APRIL -> resources.getString(R.string.mon_Apr)
            Calendar.MAY -> resources.getString(R.string.mon_May)
            Calendar.JUNE -> resources.getString(R.string.mon_June)
            Calendar.JULY -> resources.getString(R.string.mon_July)
            Calendar.AUGUST -> resources.getString(R.string.mon_Aug)
            Calendar.SEPTEMBER -> resources.getString(R.string.mon_Sep)
            Calendar.OCTOBER -> resources.getString(R.string.mon_Oct)
            Calendar.NOVEMBER -> resources.getString(R.string.mon_Nov)
            Calendar.DECEMBER -> resources.getString(R.string.mon_Dec)
            else -> ""
        }

    }

}